pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,661 @@
+}
+
+type homeWorkspaceHandler struct {
+	apiHandler http.Handler
+
+	authz authorizer.Authorizer
+
+	createClusterRole        func(ctx context.Context, workspace logicalcluster.Name, cr *rbacv1.ClusterRole) error
+	createClusterRoleBinding func(ctx context.Context, workspace logicalcluster.Name, crb *rbacv1.ClusterRoleBinding) error
+
+	createClusterWorkspace func(ctx context.Context, workspace logicalcluster.Name, cw *tenancyv1alpha1.ClusterWorkspace) error
+	getClusterWorkspace    func(ctx context.Context, workspace logicalcluster.Name, name string) (*tenancyv1alpha1.ClusterWorkspace, error)
+
+	getLocalClusterWorkspace      func(fullName logicalcluster.Name) (*tenancyv1alpha1.ClusterWorkspace, error)
+	searchLocalClusterRole        searchFunc
+	searchLocalClusterRoleBinding searchFunc
+	synced                        func() bool
+
+	homePrefix               logicalcluster.Name
+	bucketLevels, bucketSize int
+	creationDelay            int
+
+	// This is only for testing
+	mocks mockableFunctions
+}
+
+func (h *homeWorkspaceHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	if !h.synced() {
+		h.apiHandler.ServeHTTP(rw, req)
+		return
+	}
+
+	ctx := req.Context()
+	var workspaceName logicalcluster.Name
+	if lcluster, err := request.ValidClusterFrom(ctx); err != nil {
+		responsewriters.InternalError(rw, req, err)
+		return
+	} else {
+		workspaceName = lcluster.Name
+	}
+	user, ok := request.UserFrom(ctx)
+	if !ok {
+		err := errors.New("No user in HomeWorkspaces filter !")
+		responsewriters.InternalError(rw, req, err)
+		return
+	}
+	requestInfo, ok := request.RequestInfoFrom(ctx)
+	if !ok {
+		responsewriters.InternalError(rw, req, errors.New("No request Info"))
+		return
+	}
+
+	if workspaceName == tenancyv1alpha1.RootCluster &&
+		requestInfo.IsResourceRequest &&
+		requestInfo.Verb == "get" &&
+		requestInfo.APIGroup == tenancyv1beta1.SchemeGroupVersion.Group &&
+		requestInfo.Resource == "workspaces" &&
+		requestInfo.Name == "~" {
+		var homeWorkspace *tenancyv1beta1.Workspace
+		homeWorkspaceFullName := h.getHomeWorkspaceName(user.GetName())
+		homeClusterWorkspace, err := h.getLocalClusterWorkspace(homeWorkspaceFullName)
+		if err != nil {
+			responsewriters.InternalError(rw, req, err)
+			return
+		}
+		if homeClusterWorkspace != nil {
+			// The home workspace for the user exists, so let's return it to the user if he request to see its definition
+
+			homeWorkspace = &tenancyv1beta1.Workspace{}
+			projection.ProjectClusterWorkspaceToWorkspace(homeClusterWorkspace, homeWorkspace)
+		} else {
+			// Test if the user has the right to see his Home workspace even though it doesn't exists
+			// => test the get verb on the clusterworkspaces/workspace subresource named ~ in the root workspace.
+
+			attributes := homeWorkspaceAuthorizerAttributes(user, "get")
```

Comment:
why don't we authz the previous case where the home workspace was found?

Reply:
We could, but I was thinking of it this as "the right to see the definition of the home workspace when it still doesn't exist".
When the home workspace has been created with the appropriate permissions to the user, he should in any case have access to it, so should also have the right to know where it is.

Reply:
Added a comment to explain this in commit https://github.com/kcp-dev/kcp/pull/1373/commits/73e74f2f2e0057639e39be70bfe58c4c9b7b0561

Reply:
> so should also have the right to know where it is.

Am not convinced a user should.

Reply:
He can look into his local kubeconfig to get the URL.